### PR TITLE
DO NOT MERGE :: Parsing enhancements

### PR DESCRIPTION
**Pull request by kdiluca**
# Issue

The goal of this PR is to improve the time needed to parse the data for the Valhalla build process.  Will add build tile times for before and after code change.

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: lua/graph.lua [hunk 1]
```diff
@@ -1831,22 +1845,41 @@ function nodes_proc (kv, nokeys)
...
   kv["access_mask"] = bit32.bor(auto, emergency, truck, bike, foot, wheelchair, bus, hov, moped, motorcycle, taxi)
 
+  --do not store exit_info and ref if highway ~= motorway_junction
+  if kv["highway"] ~= "motorway_junction" then
+    kv["exit_to"] = nil
+    kv["ref"] = nil
+  end
+
+  --do not store name if named_jct == false and highway ~= motorway_junction
+  if (named_jct ~= true and kv["highway"] ~= "motorway_junction") then
```
Comment (gknisely) [inline lua/graph.lua]: @kdiluca Wonder if this should be a `or` here.  What do you think

Comment (kdiluca) [reply lua/graph.lua]: @gknisely  I believe we need to keep it `and` but we're still missing 10.
Master: 
```
2020/03/02 14:51:57.520537 [INFO] Number of nodes with refs (exits) = 1326
2020/03/02 14:51:57.520545 [INFO] Number of nodes with exit_to = 0
2020/03/02 14:51:57.520551 [INFO] Number of nodes with names = 127
2020/03/02 14:51:57.520558 [INFO] Number of way refs = 38536
2020/03/02 14:51:57.520564 [INFO] Number of reverse way refs = 10469
```
Code change `if (named_jct ~= true and kv["highway"] ~= "motorway_junction") then` : 
```
2020/02/28 20:59:21.630673 [INFO] Number of nodes with refs (exits) = 1326
2020/02/28 20:59:21.630685 [INFO] Number of nodes with exit_to = 0
2020/02/28 20:59:21.630696 [INFO] Number of nodes with names = 117
2020/02/28 20:59:21.630706 [INFO] Number of way refs = 38536
2020/02/28 20:59:21.630715 [INFO] Number of reverse way refs = 10469
```
Code change `if (named_jct ~= true or kv["highway"] ~= "motorway_junction") then`
```
2020/03/02 15:40:26.003763 [INFO] Number of nodes with refs (exits) = 1326
2020/03/02 15:40:26.003767 [INFO] Number of nodes with exit_to = 0
2020/03/02 15:40:26.003771 [INFO] Number of nodes with names = 0
2020/03/02 15:40:26.003775 [INFO] Number of way refs = 38536
2020/03/02 15:40:26.003779 [INFO] Number of reverse way refs = 10469
```

Comment (gknisely) [reply lua/graph.lua]: Ok.  sounds like you will have to debug this one. @kdiluca

**Comment by kdiluca** [comment]
Since the building phase is taking an extreme amount of time, I'm storing off the build and parse logs to investigate:

Build log:
timestamp | message |  
-- | -- | --
1583285437267 | building data on <test env> in region us-east-1 |  
1583285641228 | /data/valhalla/tiles / |  
1583285829631 | / |  
1583285829631 | [INFO] building the pbf data. |  
1583285829719 | 2020/03/04 01:37:09.719164 [32 | 1m[INFO][0m Start stage = build End stage = build
1583285829719 | 2020/03/04 01:37:09.719197 [32 | 1m[INFO][0m Read OSMData from temp files
1583285839970 | 2020/03/04 01:37:19.970604 [32 | 1m[INFO][0m Done
1583285839970 | 2020/03/04 01:37:19.970858 [32 | 1m[INFO][0m Creating graph edges from ways...
1583285936912 | 2020/03/04 01:38:56.912267 [32 | 1m[INFO][0m Finished with 312211324 graph edges
1583285937992 | 2020/03/04 01:38:57.991768 [32 | 1m[INFO][0m Sorting graph...
1583286191920 | 2020/03/04 01:43:11.920153 [32 | 1m[INFO][0m Finished with 238710514 graph nodes
1583286192840 | 2020/03/04 01:43:12.840221 [32 | 1m[INFO][0m Reclassifying link graph edges...
1583286250522 | 2020/03/04 01:44:10.522270 [32 | 1m[INFO][0m Class: 0 exit count = 173179
1583286250522 | 2020/03/04 01:44:10.522290 [32 | 1m[INFO][0m Class: 1 exit count = 271198
1583286250522 | 2020/03/04 01:44:10.522293 [32 | 1m[INFO][0m Class: 2 exit count = 367916
1583286250522 | 2020/03/04 01:44:10.522295 [32 | 1m[INFO][0m Class: 3 exit count = 377023
1583286250522 | 2020/03/04 01:44:10.522297 [32 | 1m[INFO][0m Class: 4 exit count = 273989
1583286250522 | 2020/03/04 01:44:10.522300 [32 | 1m[INFO][0m Class: 5 exit count = 52028
1583286250522 | 2020/03/04 01:44:10.522302 [32 | 1m[INFO][0m Class: 6 exit count = 103026
1583286250522 | 2020/03/04 01:44:10.522304 [32 | 1m[INFO][0m Class: 7 exit count = 0
1583286365510 | 2020/03/04 01:46:05.510491 [32 | 1m[INFO][0m Finished with 1327577 reclassified.  Turn channel count = 679962
1583286367212 | 2020/03/04 01:46:07.211915 [32 | 1m[INFO][0m Reclassifying ferry connection graph edges...
1583286482112 | 2020/03/04 01:48:02.111979 [32;1m[INFO][0m Finished ReclassifyFerryEdges: ferry_endpoint_count = 9305, 161666 edges reclassified. |  
1583286483358 | 2020/03/04 01:48:03.357892 [32 | 1m[INFO][0m Building 166371 tiles with 32 threads...
1583287779365 | 2020/03/04 02:09:39.365083 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 245147465
1583287779365 | 2020/03/04 02:09:39.365123 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 1536
1583287779369 | 2020/03/04 02:09:39.368975 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 226579571
1583287779369 | 2020/03/04 02:09:39.368995 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 780
1583287779386 | 2020/03/04 02:09:39.386904 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 242454795
1583287779386 | 2020/03/04 02:09:39.386926 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 1536
1583287779394 | 2020/03/04 02:09:39.394303 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 245147467
1583287779394 | 2020/03/04 02:09:39.394321 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 1536
1583287779412 | 2020/03/04 02:09:39.412484 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 245147466
1583287779412 | 2020/03/04 02:09:39.412509 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 1542
1583287779416 | 2020/03/04 02:09:39.416841 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 245147309
1583287779416 | 2020/03/04 02:09:39.416859 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 1538
1583287779427 | 2020/03/04 02:09:39.427747 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199487944
1583287779427 | 2020/03/04 02:09:39.427766 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 771
1583287779483 | 2020/03/04 02:09:39.483141 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 242681999
1583287779483 | 2020/03/04 02:09:39.483577 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 772
1583287779487 | 2020/03/04 02:09:39.487323 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 619367507
1583287779487 | 2020/03/04 02:09:39.487350 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 776
1583287779488 | 2020/03/04 02:09:39.488810 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 245403943
1583287779488 | 2020/03/04 02:09:39.488825 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 776
1583287779490 | 2020/03/04 02:09:39.490280 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 418238119
1583287779490 | 2020/03/04 02:09:39.490292 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 772
1583287779496 | 2020/03/04 02:09:39.496878 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 226396749
1583287779496 | 2020/03/04 02:09:39.496901 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 772
1583287779499 | 2020/03/04 02:09:39.499118 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 245018535
1583287779499 | 2020/03/04 02:09:39.499136 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 776
1583287779575 | 2020/03/04 02:09:39.575858 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199429789
1583287779575 | 2020/03/04 02:09:39.575886 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 772
1583287779597 | 2020/03/04 02:09:39.597134 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199490089
1583287779597 | 2020/03/04 02:09:39.597170 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 769
1583287779787 | 2020/03/04 02:09:39.787226 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 619367509
1583287779787 | 2020/03/04 02:09:39.787259 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 769
1583287779802 | 2020/03/04 02:09:39.802049 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 228683941
1583287779802 | 2020/03/04 02:09:39.802068 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 770
1583287780019 | 2020/03/04 02:09:40.019357 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 228683948
1583287780019 | 2020/03/04 02:09:40.019384 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 770
1583287780027 | 2020/03/04 02:09:40.027090 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 675577898
1583287780027 | 2020/03/04 02:09:40.027112 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 257
1583287780030 | 2020/03/04 02:09:40.030768 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 242832577
1583287780030 | 2020/03/04 02:09:40.030785 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 769
1583287780031 | 2020/03/04 02:09:40.031505 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 587698764
1583287780031 | 2020/03/04 02:09:40.031519 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 514
1583287780122 | 2020/03/04 02:09:40.122212 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 675577856
1583287780122 | 2020/03/04 02:09:40.122237 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 770
1583287780154 | 2020/03/04 02:09:40.154037 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 226562982
1583287780154 | 2020/03/04 02:09:40.154056 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 385
1583287780205 | 2020/03/04 02:09:40.205180 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 243905601
1583287780205 | 2020/03/04 02:09:40.205205 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 386
1583287780230 | 2020/03/04 02:09:40.230144 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 675577859
1583287780230 | 2020/03/04 02:09:40.230172 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 769
1583287780315 | 2020/03/04 02:09:40.315785 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 389861635
1583287780315 | 2020/03/04 02:09:40.315814 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 770
1583287780320 | 2020/03/04 02:09:40.320647 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 242832553
1583287780320 | 2020/03/04 02:09:40.320668 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 770
1583287780449 | 2020/03/04 02:09:40.449371 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 228682173
1583287780449 | 2020/03/04 02:09:40.449396 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 768
1583287780450 | 2020/03/04 02:09:40.450875 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 228683923
1583287780450 | 2020/03/04 02:09:40.450894 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 520
1583287780465 | 2020/03/04 02:09:40.465706 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199430659
1583287780465 | 2020/03/04 02:09:40.465724 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 776
1583287780473 | 2020/03/04 02:09:40.473249 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199430660
1583287780473 | 2020/03/04 02:09:40.473264 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 769
1583287780597 | 2020/03/04 02:09:40.597228 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 389894999
1583287780597 | 2020/03/04 02:09:40.597260 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 1536
1583287781561 | 2020/03/04 02:09:41.561097 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199430314
1583287781561 | 2020/03/04 02:09:41.561124 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 772
1583287781561 | 2020/03/04 02:09:41.561876 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 242832559
1583287781561 | 2020/03/04 02:09:41.561891 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 770
1583287781562 | 2020/03/04 02:09:41.562642 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 199430661
1583287781562 | 2020/03/04 02:09:41.562657 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 776
1583287781574 | 2020/03/04 02:09:41.574300 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 226562983
1583287781574 | 2020/03/04 02:09:41.574322 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 772
1583287781591 | 2020/03/04 02:09:41.591695 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 242832546
1583287781591 | 2020/03/04 02:09:41.591717 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 768
1583287781592 | 2020/03/04 02:09:41.592439 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit on wayid: 243878466
1583287781592 | 2020/03/04 02:09:41.592457 [33 | 1m[WARN][0m Restrictions mask exceeds allowable limit: 388


Parse log:
timestamp | message |  
-- | -- | --
1583195214723 | parsing data on <test env> in region us-east-1 |  
1583195214724 | [INFO] updating config. |  
1583195214757 | [INFO] downloading <data>.pbf |  
1583195434662 | [INFO] successfully downloaded . |  
1583195435343 | [INFO] downloading  <data>.sqlite. |  
1583195437678 | [INFO] downloading <data>.sqlite |  
1583195438917 | [INFO] parsing the pbf data. |  
1583195438965 | 2020/03/03 00:30:38.965481 [32 | 1m[INFO][0m Start stage = parse End stage = parse
1583195439908 | 2020/03/03 00:30:39.908698 [32 | 1m[INFO][0m Parsing files: /data/valhalla/extracts/1583195214.pbf
1583195439908 | 2020/03/03 00:30:39.908907 [32 | 1m[INFO][0m Parsing ways...
1583196166051 | 2020/03/03 00:42:46.050811 [33;1m[WARN][0m Max osmid exceeded bitset, resizing to fit id: 6951903904 |  
1583196166699 | 2020/03/03 00:42:46.699430 [33;1m[WARN][0m Max osmid exceeded bitset, resizing to fit id: 7089449887 |  
1583196166733 | 2020/03/03 00:42:46.733360 [33;1m[WARN][0m Max osmid exceeded bitset, resizing to fit id: 7208084769 |  
1583196166832 | 2020/03/03 00:42:46.831935 [33;1m[WARN][0m Max osmid exceeded bitset, resizing to fit id: 6990870314 |  
1583196167499 | 2020/03/03 00:42:47.499373 [33;1m[WARN][0m Max osmid exceeded bitset, resizing to fit id: 7079648541 |  
1583196167539 | 2020/03/03 00:42:47.539291 [33;1m[WARN][0m Max osmid exceeded bitset, resizing to fit id: 7223004840 |  
1583205082530 | 2020/03/03 03:11:22.529851 [32 | 1m[INFO][0m invalid_argument thrown for condition (Jan 01- May 31)
1583205082530 | 2020/03/03 03:11:22.530778 [32 | 1m[INFO][0m invalid_argument thrown for condition (Jan 01- May 31)
1583208528613 | 2020/03/03 04:08:48.613361 [32 | 1m[INFO][0m invalid_argument thrown for condition (Jun 15-Labor Day)
1583209705388 | 2020/03/03 04:28:25.388757 [32 | 1m[INFO][0m invalid_argument thrown for condition (Jun 15-Labor Day)
1583210796956 | 2020/03/03 04:46:36.955965 [32 | 1m[INFO][0m invalid_argument thrown for condition (Jan 01- May 31)
1583219781495 | 2020/03/03 07:16:21.495298 [32 | 1m[INFO][0m Finished with 146217434 routable ways containing 1666435440 nodes
1583219781495 | 2020/03/03 07:16:21.495319 [32 | 1m[INFO][0m Sorting osm access tags by way id...
1583219781891 | 2020/03/03 07:16:21.890862 [32 | 1m[INFO][0m Parsing relations...
1583221631974 | 2020/03/03 07:47:11.974012 [32 | 1m[INFO][0m invalid_argument thrown for condition  Jul 1- Aug 31 off
1583221641121 | 2020/03/03 07:47:21.121052 [32 | 1m[INFO][0m Finished with 1195148 simple restrictions
1583221641121 | 2020/03/03 07:47:21.121071 [32 | 1m[INFO][0m Finished with 0 lane connections
1583221641192 | 2020/03/03 07:47:21.192503 [32 | 1m[INFO][0m Sorting complex restrictions by from id...
1583221641253 | 2020/03/03 07:47:21.253043 [32 | 1m[INFO][0m Sorting complex restrictions by to id...
1583221641310 | 2020/03/03 07:47:21.309970 [32 | 1m[INFO][0m Sorting osm way node references by node id...
1583222255682 | 2020/03/03 07:57:35.682041 [32 | 1m[INFO][0m Finished
1583222255682 | 2020/03/03 07:57:35.682067 [32 | 1m[INFO][0m Parsing nodes...
1583233128212 | 2020/03/03 10:58:48.211852 [32 | 1m[INFO][0m Finished with 1446717196 nodes contained in routable ways
1583233128212 | 2020/03/03 10:58:48.212112 [32 | 1m[INFO][0m Sorting osm way node references by way index and node shape index...
1583233733717 | 2020/03/03 11:08:53.717015 [32 | 1m[INFO][0m Finished: changeset id 81522556
1583233733717 | 2020/03/03 11:08:53.717035 [32 | 1m[INFO][0m Number of nodes with refs (exits) = 99685
1583233733717 | 2020/03/03 11:08:53.717041 [32 | 1m[INFO][0m Number of nodes with exit_to = 13252
1583233733717 | 2020/03/03 11:08:53.717045 [32 | 1m[INFO][0m Number of nodes with names = 130550
1583233733717 | 2020/03/03 11:08:53.717050 [32 | 1m[INFO][0m Number of way refs = 247738
1583233733717 | 2020/03/03 11:08:53.717054 [32 | 1m[INFO][0m Number of reverse way refs = 21021
1583233733717 | 2020/03/03 11:08:53.717058 [32;1m[INFO][0m Unique Node Strings (names, refs, etc.) = 97249 |  
1583233733717 | 2020/03/03 11:08:53.717063 [32;1m[INFO][0m Unique Strings (names, refs, etc.) = 9098879 |  
1583233733794 | 2020/03/03 11:08:53.794140 [32 | 1m[INFO][0m Write OSMData to temp files
1583233737540 | 2020/03/03 11:08:57.540068 [32 | 1m[INFO][0m Done
1583233744315 | /data/valhalla/tiles /data/valhalla |  
1583233815240 | /data/valhalla |  
1583233815240 | /data/valhalla/parse_2020_03_01-03_00_00 /data/valhalla |  
1583233816197 | [INFO] uploading admin_2020_03_01-03_00_00.sqlite |  
1583233819701 | [INFO] uploading parse_2020_03_01-03_00_00.tar |  
1583234238246 | [INFO] uploading tz_world_2020_03_01-03_00_00.sqlite |  
1583234239636 | [SUCCESS] parsing complete |

**Comment by kdiluca** [comment]
After testing all changes to pbfgraphparser.cc and graph.lua (for performance enhancements), against PA, MD (for driveways) and Switzerland(for sac_scale & mtb)...  we felt confident that all was good because all counts matched up against master parsing runs.  

We deployed the refactor branch to guidance-staging with updated osm data.  Rechecked the #'s this morning against the parsing log #'s before the refactor.  Everything matched except for:
```
Before Refactor: Number of nodes with names = 130550 
vs. 
After Refactor: Number of nodes with names = 114728
Before Refactor: Unique Node Strings (names, refs, etc.) = 97249 
vs. 
After Refactor: Unique Node Strings (names, refs, etc.) = 83647
```
This isn't a huge concern but does show us that we missed an edge case.

In summary, the performance times of the refactor did NOT speed things up which is surprising and unfortunate. It still takes approx. 11 hours.

**Next step:** We are updating the stack to increase the memory and cpu to test performance.
